### PR TITLE
Temporary password resets

### PR DESCRIPTION
**Pull request by aaronFranssell**
Ticket: https://securedatacommons.atlassian.net/browse/SDC-1743
Related PR: https://github.com/usdot-jpo-sdc-projects/sdc-dot-cognito-account-sync/pull/32

Unfortunately some of the code I needed was already in master, so I did a git pull...

This PR adds screens + logic around temporary password resets.

It also adds Login.gov logins back into the develop branch.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -15,7 +15,6 @@ matrix:
         - TEST_DIR=webportal/webapp/
       script:
         - cd $TEST_DIR && npm install && npm run test-headless
-        - cd ../../ && sonar-scanner -X -D sonar.login=$SONAR_TOKEN
```
Comment (aaronFranssell) [inline .travis.yml:18]: This got pulled in from master.

File: sonar-project.properties [hunk 2]
```diff
@@ -12,7 +12,7 @@ sonar.language=py
 sonar.sources=webportal/lambda/deployment
 sonar.tests=webportal/lambda/tests/deployment
 sonar.exclusions=*coverage*.xml, **/__init__.py
-sonar.python.coverage.reportPaths = *coverage*.xml
```
Comment (aaronFranssell) [inline sonar-project.properties:15]: This got pulled in from master.

File: webportal/webapp/angular.json [hunk 3]
```diff
@@ -53,6 +53,23 @@
                   "with": "src/environments/environment.dev-private.ts"
                 }
               ]
+            },
```
Comment (aaronFranssell) [inline webportal/webapp/angular.json:56]: This got pulled in from master. It adds a test environment.

File: webportal/webapp/src/app/account/loginsync/components/loginsync.component.html [hunk 4]
```diff
@@ -1,24 +1,47 @@
...
-      <form class="usa-form signin" (ngSubmit)="onSubmit()">
-        <fieldset class="usa-fieldset">
-          <legend class="usa-legend">One Time Sign In</legend>
-          <label for="username">Username</label>
-          <input class="usa-input" id="username" name="username" [(ngModel)]="username" autocapitalize="off" autocorrect="off">
-  
-          <label for="password-sign-in">Password</label>
-          <input class="usa-input" id="password-sign-in" name="password" [(ngModel)]="password" type="password" appTogglePassword>
-          <input class="usa-button link-submit" type="submit" value="Sign in">
+    <div class="grid-row margin-top-3">
```
Comment (aaronFranssell) [inline webportal/webapp/src/app/account/loginsync/components/loginsync.component.html:3]: I merged the regular sync account screen with the password reset screen. Most of the functionality is shared, and they seem to fit pretty well together. The component + view don't seem to be too large...

File: webportal/webapp/src/app/account/loginsync/components/loginsync.component.ts [hunk 5]
```diff
@@ -11,34 +12,73 @@ import { CognitoService } from '../../../../services/cognito.service';
...
-  linkSuccessful = true;
+  newPassword: string;
+  newPasswordConfirmation: string;
+  changeTemporaryPassword = false;
   errorMessage = '';
+  complexityErrorMessage: string = "The password does not match the complexity criteria.";
+  passwordMatchingErrorMessage: string = "Passwords must match.";
 
-  constructor(private loginSyncService: LoginSyncService, private cognitoService: CognitoService) { }
+  constructor(private loginSyncService: LoginSyncService, private cognitoService: CognitoService,  @Inject(WindowToken) private window: Window) { }
```
Comment (aaronFranssell) [inline webportal/webapp/src/app/account/loginsync/components/loginsync.component.ts:22]: This window token is how Angular suggests injecting `window` into components for testing.

File: webportal/webapp/src/app/account/loginsync/components/loginsync.component.spec.ts [hunk 6]
```diff
@@ -1,25 +1,172 @@
...
+
+                expect(mockLoginSyncService.resetTemporaryPassword).toHaveBeenCalledTimes(0);
+                expect(fixture.debugElement.nativeElement.querySelector('#alert_show_alert').textContent).toBeTruthy();
+                expect(fixture.debugElement.nativeElement.querySelector('#alert_show_alert').textContent).toEqual(testCase.error);
+             });
+        });
+
+        it('it shows a nice error message if something bad happens during the password reset', () => {
+            spyOn(mockLoginSyncService, 'resetTemporaryPassword').and.returnValue(Observable.throw({ userErrorMessage: 'Oh noes...', body: { } }));
+            component.newPassword = 'Lets-Switch-To-React';
```
Comment (aaronFranssell) [inline webportal/webapp/src/app/account/loginsync/components/loginsync.component.spec.ts:100]: Not a super fan of Angular...

File: webportal/webapp/src/app/account/loginsync/components/loginsync.component.html [hunk 7]
```diff
@@ -1,24 +1,47 @@
...
+
+            <p class="margin-top-4 padding-top-4 border-top">Your temporary Secure Data Commons password must be changed.</p>
+            <p>Your new password must be 7 characters long and contain at least 3 of the following:</p>
+            <ul>
+              <li>Lower case letter</li>
+              <li>Upper case letter</li>
+              <li>Special character</li>
+              <li>Number</li>
+            </ul>
+            <p>Your password may not be the same as your previous 23 passwords.</p>
```
Comment (aaronFranssell) [inline webportal/webapp/src/app/account/loginsync/components/loginsync.component.html:41]: Very specific number...

File: webportal/webapp/src/app/account/loginsync/guards/loginsync.guard.spec.ts [hunk 8]
```diff
@@ -1,31 +1,100 @@
 import { LoginSyncGuard } from './loginsync.guard';
+import { CognitoService } from '../../../../services/cognito.service';
```
Comment (aaronFranssell) [inline webportal/webapp/src/app/account/loginsync/guards/loginsync.guard.spec.ts:2]: Boy scout rule... my changes were minor to the actual file, but I added tests for it.

File: webportal/webapp/src/app/app.module.ts [hunk 9]
```diff
@@ -13,6 +13,7 @@ import { MatTableModule } from '@angular/material/table';
 import { CdkTableModule } from '@angular/cdk/table';
 import { TableModule } from 'primeng/table';
 import { FileUploadModule } from 'primeng/fileupload';
+import { WindowToken, windowProvider } from '../factories/window.factory';
```
Comment (aaronFranssell) [inline webportal/webapp/src/app/app.module.ts:16]: This all got pulled in from master...

File: webportal/webapp/src/app/main/register/register.component.ts [hunk 10]
```diff
@@ -17,14 +17,7 @@ export class RegisterComponent implements OnInit {
     if (this.email.endsWith('dot.gov')) {
       this.cognitoService.login(false); // Route to ADFS login
     } else {
-      window.location.href = this.buildLoginGovUrl(); // Route to Login.gov
+      window.location.href = this.cognitoService.buildLoginGovRedirectUrl();
     }
   }
-
-  buildLoginGovUrl() {
```
Comment (aaronFranssell) [inline webportal/webapp/src/app/main/register/register.component.ts:24]: This got pulled in from master, but the idea is that this should be encapsulated

File: webportal/webapp/src/environments/environment.prod.ts [hunk 11]
```diff
@@ -3,14 +3,16 @@ export const environment = {
   production: true,
   STREAMING_URL: 'https://stream.securedatacommons.com/guacamole/?authToken=',
   REGION: 'us-east-1',
-  USER_POOL_ID: 'us-east-1_Y5JI7ysvY',
-  CLIENT_ID: '7t034p7j0aklgipi5kk8ll5udd',
+  USER_POOL_ID: 'us-east-1_1zLqP2XrI',
```
Comment (aaronFranssell) [inline webportal/webapp/src/environments/environment.prod.ts:6]: Pulled in from master... this points us to the correct Cognito user pool.

File: webportal/webapp/src/environments/environment.prod.ts [hunk 12]
```diff
@@ -3,14 +3,16 @@ export const environment = {
...
   REDIRECT_URL: 'https://portal.securedatacommons.com/index.html',
-  API_ENDPOINT: 'https://cyfca70occ.execute-api.us-east-1.amazonaws.com/api/',
-  LOGIN_GOV_COGNITO_APP_CLIENT_ID: '34vj4387sna9jk31gudpt3b78i',
+  API_ENDPOINT: 'https://bzzlrrsxa4.execute-api.us-east-1.amazonaws.com/api/',
+  LOGIN_GOV_COGNITO_APP_CLIENT_ID: '3mlku34f58oeloqqc3hpiko6l3',
   ACCOUNT_LINK_URL: 'https://2l418womp5.execute-api.us-east-1.amazonaws.com/prod',
   LINK_ACCOUNT_PATH: 'prod-link-account',
   ACCOUNT_LINKED_PATH: 'prod-account-linked',
+  RESET_TEMPORARY_PASSWORD_PATH: 'prod-reset-temporary-password',
+  ENVIRONMENT: 'prod',
```
Comment (aaronFranssell) [inline webportal/webapp/src/environments/environment.prod.ts:17]: Justin, do we need this `ENVIRONMENT` anymore? Its superseded by `ACCOUNT_LINK_URL` right?

File: webportal/webapp/src/environments/environment.test.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
+// The file contents for the current environment will overwrite these during build.
+// The build system defaults to the dev environment which uses `environment.ts`, but if you do
+// `ng build --env=prod` then `environment.prod.ts` will be used instead.
+// The list of which env maps to which file can be found in `.angular-cli.json`.
+
+export const environment = {
```
Comment (aaronFranssell) [inline webportal/webapp/src/environments/environment.test.ts:6]: This got pulled in from master (and I tweaked it to work with the new Nginx proxy naming conventions).

File: webportal/webapp/src/factories/window.factory.ts [hunk 14]
```diff
@@ -0,0 +1,4 @@
+import { InjectionToken } from '@angular/core';
```
Comment (aaronFranssell) [inline webportal/webapp/src/factories/window.factory.ts:1]: Pulled in from master.

File: webportal/webapp/src/services/cognito.service.ts [hunk 15]
```diff
@@ -28,33 +29,24 @@ export interface Callback {
...
     public static _IDENTITY_PROVIDER = environment.IDENTITY_PROVIDER // User pool Identity provider name
     public static _APP_DOMAIN = environment.APP_DOMAIN // App domain name
     public static _IDP_ENDPOINT = "cognito-idp." + CognitoService._REGION + ".amazonaws.com/" + CognitoService._USER_POOL_ID
-    public static _REDIRECT_URL = environment.REDIRECT_URL // Re-direct URL for the user pool
     
     public static _POOL_DATA:any = {
         UserPoolId: CognitoService._USER_POOL_ID,
         ClientId: CognitoService._CLIENT_ID
     };
-    public static _AUTH_DATA = {
```
Comment (aaronFranssell) [inline webportal/webapp/src/services/cognito.service.ts:44]: This got pulled in from master... this had to get refactored slightly to use the injected window.

File: webportal/webapp/test_deploy.sh [hunk 16]
```diff
@@ -0,0 +1,6 @@
+#/bin/bash
```
Comment (aaronFranssell) [inline webportal/webapp/test_deploy.sh:1]: Pulled in from master...

File: webportal/webapp/src/app/account/loginsync/components/loginsync.component.spec.ts (hunk 6 above)
Comment (TiannaW24) [reply webportal/webapp/src/app/account/loginsync/components/loginsync.component.spec.ts:100]: 😆

File: webportal/webapp/src/environments/environment.prod.ts (hunk 12 above)
Comment (JAMSUPREME) [reply webportal/webapp/src/environments/environment.prod.ts:17]: Yeah. Wouldn't need it for URL generation. It'd only be useful for informational purposes or doing some sorta logic based on environment _(though doing that sorta thing on the front-end probably isn't ideal)_

**Pull request merged**